zerver/lib/push_notifications.py

```diff
@@ -787,22 +869,50 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
     gcm_payload, gcm_options = get_message_payload_gcm(user_profile, message)
     logger.info("Sending push notifications to mobile clients for user %s", user_profile_id)
 
-    if uses_notification_bouncer():
-        send_notifications_to_bouncer(user_profile_id,
-                                      apns_payload,
-                                      gcm_payload,
-                                      gcm_options)
-        return
-
     android_devices = list(PushDeviceToken.objects.filter(user=user_profile,
                                                           kind=PushDeviceToken.GCM))
 
     apple_devices = list(PushDeviceToken.objects.filter(user=user_profile,
                                                         kind=PushDeviceToken.APNS))
 
+    encrypted_android_devices, android_devices = separate_devices(android_devices)
+
+    encrypted_apple_devices, apple_devices = separate_devices(apple_devices)
+
+    encrypted_gcm_data = generate_encrypted_data(encrypted_android_devices,
+                                                 gcm_payload,
+                                                 encrypt_gcm_payload)
+
+    encrypted_apns_data = generate_encrypted_data(encrypted_apple_devices,
+                                                  apns_payload,
+                                                  encrypt_apns_payload)
+
+    if uses_notification_bouncer():
+        if android_devices or apple_devices:
+            send_notifications_to_bouncer(user_profile_id,
+                                          apns_payload,
+                                          gcm_payload,
+                                          gcm_options)
+
+        if encrypted_android_devices or encrypted_apple_devices:
+            encrypted_apns_payload = prepare_encrypted_payload(encrypted_apns_data)
+            encrypted_gcm_payload = prepare_encrypted_payload(encrypted_gcm_data)
+            send_notifications_to_bouncer(user_profile_id,
+                                          encrypted_apns_payload,
+                                          encrypted_gcm_payload,
+                                          gcm_options,
+                                          encrypted=True)
+        return
+
     if apple_devices:
         send_apple_push_notification(user_profile.id, apple_devices,
                                      apns_payload)
 
     if android_devices:
         send_android_push_notification(android_devices, gcm_payload, gcm_options)
+
+    if encrypted_apple_devices:
+        send_apple_encrypted_push_notification(user_profile.id, encrypted_apns_data)
```

Comment:
The `if` here has no effect at all, and just makes the code more complicated.

The existing conditionals `if apple_devices` and `if android_devices` should really be removed too -- that would make both this code and the zilencer view simpler, and would help with the new complexity that this PR has to add. If you *just* remove them I think that will cause trouble, but it's easy to fix: each of these functions should just make sure that if it's given an empty list of devices, it does nothing.

That would be good to do as a small commit at the start of the branch.